File: python/blocking_error.py
import dcss_api
from dcss_api import BlockingErr, APIErr

print(dcss_api.WebtilePy.read_until.__doc__)


# # Connect to DCSS Webtile
# webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 100, "0.29")

# # Print the messages you get upon connecting
# while (message := webtile.get_message()) != None:
#     print(message)

# # Log in (to a user called "Username", with a password "Password")
# webtile.login_with_credentials("Username", "Password")

# # Print the messages you get upon connecting
# while (message := webtile.get_message()) != None:
#     print(message)

# # Start game
# webtile.write_json("""{"msg": "play", "game_id": "seeded-web-trunk"}""")
# webtile.read_until("player", None, None)
# webtile.write_key("-")
# webtile.read_until("ui-state-sync", None, None)
# webtile.write_key("1")
# webtile.write_key("\t\t\t \r")
# webtile.read_until("ui-push", None, None)
# webtile.write_key("b")
# webtile.write_key("i")
# webtile.write_key("b")
# webtile.read_until("input_mode", "mode", 1)

# webtile.write_key("6")
# webtile.write_key("i")
# webtile.write_key("a")
# webtile.write_key("d")
# webtile.write_key("i")
# webtile.write_key("b")
# webtile.write_key("d")
# webtile.write_key(",")

# try:
#     webtile.read_until("ignore", None, None)
# except BlockingErr as e:
#     if e.args[0] == "Pickup":
#         print(f"Blocking (Pickup): {e}")
#     else:
#         print(f"Blocking: {e}")
# except APIErr as e:
#     print(f"API Error: {e}")

File: examples/blocking_error.rs
extern crate dcss_api;

use dcss_api::Webtile;
use dcss_api::{BlockingError, Error};
use serde_json::json;

fn main() {
    // Connect to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 100, "0.29").expect("Failed to connect");

    // Print the messages you get upon connecting
    while let Some(message) = webtile.get_message() {
        println!("{:?}", message)
    }

    // Log in (to a user called "Username", with a password "Password")
    let _gameid = webtile
        .login_with_credentials("Username", "Password")
        .expect("Failed to login");

    // Print the messages you get upon connecting
    while let Some(message) = webtile.get_message() {
        println!("{:?}", message)
    }

    // Start game
    webtile
        .write_json(json!({"msg": "play", "game_id": "seeded-web-trunk"}))
        .expect("");
    webtile.read_until("player", None, None).expect("");
    webtile.write_key("-").expect("");
    webtile.read_until("ui-state-sync", None, None).expect("");
    webtile.write_key("1").expect("");
    webtile.write_key("\t\t\t \r").expect("");
    webtile.read_until("ui-push", None, None).expect("");
    webtile.write_key("b").expect("");
    webtile.write_key("i").expect("");
    webtile.write_key("b").expect("");
    webtile
        .read_until("input_mode", Some("mode"), Some(1))
        .expect("");

    webtile.write_key("6").expect("");
    webtile.write_key("i").expect("");
    webtile.write_key("a").expect("");
    webtile.write_key("d").expect("");
    webtile.write_key("i").expect("");
    webtile.write_key("b").expect("");
    webtile.write_key("d").expect("");
    webtile.write_key(",").expect("");

    match webtile.read_until("ignore", None, None) {
        Ok(_) => (),
        Err(e) => match e {
            Error::Blocking(BlockingError::Pickup) => print!("PICKUP"),
            Error::Blocking(BlockingError::Died) => print!("PICKUP"),
            _ => panic!("Unexpected Error"),
        },
    };
}
